### PR TITLE
Fix open_dataset regression

### LINKED ISSUE #5098: open_dataset regression
**What happened**:

`open_dataset` doesn't replace `~` any more

**What you expected to happen**:

should replace `~` for backward compatibility

**Minimal Complete Verifiable Example**:

```python
import xarray as xr

da = xr.DataArray([1, 2, 3])
da.to_netcdf("~/bug_report.nc")

xr.open_dataarray("~/bug_report.nc")
```

**Anything else we need to know?**:
works on 0.17.0, fails on master

### PR DESCRIPTION
**Pull request by aurghs**
Fix `open_dataset` regression, expands ~ in `filepath_or_obj` when necessary. 
I have checked the behaviour of the engines. It seems that `pynio`  already expands  ~.

- [x] Closes #5098
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by bcbnz** [comment]
I think this would also fix #5132 if `os.path.abspath` was added around the `os.path.expanduser` calls.

**Comment by dcherian** [comment]
LGTM. I don't know how we would test this...

**Comment by alexamici** [comment]
@aurghs & @dcherian note that at the moment most backends accept `pathlib.Path` (but it is not tested), so the `isintance(..., str)` doesn't run `expandpath` in that case.

**Comment by aurghs** [comment]
> LGTM. I don't know how we would test this...

I don't have any good idea about it.

File: xarray/backends/pseudonetcdf_.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+import os
```
Comment (alexamici) [inline xarray/backends/pseudonetcdf_.py:1]: Please do not import all `os`
```suggestion
import os.path
```

**Comment by aurghs** [comment]
> @aurghs & @dcherian note that at the moment most backends accept `pathlib.Path` (but it is not tested), so the `isintance(..., str)` doesn't run `expandpath` in that case.
What do you suggest?

**Comment by bcbnz** [comment]
> > @aurghs & @dcherian note that at the moment most backends accept `pathlib.Path` (but it is not tested), so the `isintance(..., str)` doesn't run `expandpath` in that case.
> 
> What do you suggest?

For Python 3.6+ expanduser works with paths, returning a string:

```
>>> os.path.expanduser(pathlib.Path("~/file.nc"))
'/home/username/file.nc'
```
If its preferred to keep a Path instance, then the [Path.expanduser](https://docs.python.org/3/library/pathlib.html#pathlib.Path.expanduser) method is the equivalent.

**Review by alexamici — COMMENTED** [review]
@aurghs you may try:

```python
try:
    filename_or_obj = os.path.expanduser(filename_or_obj)
except TypeError:
    pass
```

**Comment by alexamici** [comment]
@aurghs you may also use `_normalize_path` that does all the checks:

https://github.com/pydata/xarray/blob/master/xarray/backends/api.py#L112

**Comment by bcbnz** [comment]
> LGTM. I don't know how we would test this...


For ensuring absolute paths, my MVCE in #5132 could be adapted.

Expansion of ~ might be trickier. The [os.path.expanduser](https://docs.python.org/3/library/os.path.html#os.path.expanduser) docstring says

> On Unix, an initial ~ is replaced by the environment variable HOME if it is set

and

> On Windows, USERPROFILE will be used if set

so tests run with appropriate environment variables could be used with OS checking.

**Comment by aurghs** [comment]
At this point, probably `api.normalize_path` is the best choice.

**Review by alexamici — APPROVED** [review]
LGTM

**Comment by alexamici** [comment]
@aurghs can you confirm that all failures are unrelated to the changes?

**Comment by keewis** [comment]
@alexamici, that's #5147

**Comment by aurghs** [comment]
> @aurghs can you confirm that all failures are unrelated to the changes?

In my understanding, the errors are not related to the changes.

**Comment by aurghs** [comment]
@bcbnz could you check if this fixes also #5132?

**Comment by bcbnz** [comment]
@aurghs it fixes it for both the test script in #5132 (using all three netCDF engines, ``netcdf4``, ``h5netcdf`` and ``scipy``) and for my original unit tests where I found the problem.

**Comment by alexamici** [comment]
@bcbnz thanks for the help! I'll merge it.

**Pull request merged**